# NoahCare Structural Variant Calling

Ncsv is a structural variant calling, annotation and filting software. It uses Split Reads, Single Unmap and Discordant Pair to discover the structural variants throughout the genome.

## Installing ncsv

```
git clone https://github.com/GenePlus/ncsv.git --recursive
sh ncsv/inst.sh #compile (cd ncsv/extension/sv_helper && python setup.py build_ext --inplace #generate so file)
```

## Getting Started

Ncsv use a run_svp.py(https://github.com/GenePlus/ncsv/blob/master/bin/run_sv.py) file as a start step.
```
python2 bin/run_sv.py [-h] 
        [-S SAMPLE] -t TMP -r REF -H HOTREGION
        [-w BWA] [-s SAMTOOLS] [-n NUMTHREADS]
        [-d] [--bamstat BAMSTAT] [--hotsv HOTSV]
        [-c CTRLOUT] outfile bamt [bamn]
    
positional arguments:

    outfile                 output file, csv format
    bamt                    bam to do sv analysis
    bamn                    ctrl bam

optional arguments:

    -h, --help                              show this help message and exit
    -t TMP, --tmp TMP                       tmp file path
    -r REF, --ref REF                       reference genome, it should be same with the one used by bam
    -H HOTREGION, --hotregion HOTREGION     hot region file, generated by sv flow helper script
    -w BWA, --bwa BWA                       default is bwa
    -s SAMTOOLS, --samtools SAMTOOLS        default is samtools
    -n NUMTHREADS, --numthreads NUMTHREADS  num of threads, default is 4
    -c CTRLOUT, --ctrlout CTRLOUT           norm sv output file, csv format
    -S SAMPLE, --sample SAMPLE              sample name, set None to infer it from bam path
    -d, --debug                             open debug logging
    --bamstat BAMSTAT                       bam stat file, isize could be abstracted from it
    --hotsv HOTSV                           the bed file as hot sv region. WARNING: only analysis the sv inner these region
```

More docs refers to [ncsv](https://ncsv.readthedocs.io/en/latest/index.html) on read the docs.

## License

ncsv source code is provided under the [Apache License 2.0](https://github.com/GenePlus/ncsv/blob/master/LICENSE).
ncsv includes several third party packages provided under other open source licenses, please see [COPYRIGHT.txt](https://github.com/GenePlus/ncsv/blob/master/COPYRIGHT.txt) for additional details. Specially, this program use [svaba](https://github.com/walaj/svaba) to process local assembly.




